### PR TITLE
🐛 Skip updating VMOp immutable fields

### LINKED ISSUE #2430: Remove immutable fields when patching VM Service VMs after creation
/kind feature

**Describe the solution you'd like**
Currently, the following fields are immutable after a VM Service VM resource is created ([related code in vmop repo](https://github.com/vmware-tanzu/vm-operator/blob/main/webhooks/virtualmachine/v1alpha1/validation/virtualmachine_validator.go#L143-L148)):
- ImageName
- ClassName
- StorageClass
- ResourcePolicyName

Therefore, patching these fields after a VM Service VM is created could potentially be denied, as they might be updated during the VM creation by a mutation webhook.

**Anything else you would like to add:**
For example, in the latest VMOP version, `vm.spec.imageName` can be mutated from a friendly image name to a "vmi-" name if there is a single VM image that can be resolved. If CAPV continues patching such VM with the previous friendly image name, the request would be denied as the imageName is an immutable field, leading to reconciliation errors for TKG VMs.

**Environment:**
N/A.

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
VMOp mutation webhook updates the VMOp Spec.ImageName to a user-friendly name prefixed `vmi-` or `cvmi-`. This causes the reconciliation of the VSphereMachine to fail since the controllers copy the all spec fields to the VMOp `VirtualMachine` on every reconcile. Since imageName is an immutable field we see the below error:
```yaml
Name:         guestcluster1-control-plane-bz7k9-c9whq
Namespace:    gctest
Labels:       cluster.x-k8s.io/cluster-name=guestcluster1
              cluster.x-k8s.io/control-plane=
              cluster.x-k8s.io/control-plane-name=guestcluster1-control-plane
              topology.cluster.x-k8s.io/owned=
Annotations:  cluster.x-k8s.io/cloned-from-groupkind: VSphereMachineTemplate.vmware.infrastructure.cluster.x-k8s.io
              cluster.x-k8s.io/cloned-from-name: guestcluster1-control-plane-bz7k9
              run.tanzu.vmware.com/resolve-os-image: os-name=photon
API Version:  vmware.infrastructure.cluster.x-k8s.io/v1beta1
Kind:         VSphereMachine
Metadata:
  Creation Timestamp:  2023-12-14T20:01:42Z
  Finalizers:
    vspheremachine.infrastructure.cluster.x-k8s.io
  Generation:  2
  Owner References:
    API Version:           cluster.x-k8s.io/v1beta1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Machine
    Name:                  guestcluster1-control-plane-wr5w2
    UID:                   fd9da4eb-8b09-4484-9e1a-90a284861f4c
  Resource Version:        415153
  UID:                     1ba496ef-ccc4-4a6a-9e29-4a8009ec7957
Spec:
  Class Name:      best-effort-xsmall
  Failure Domain:  vmware-system-legacy
  Image Name:      tkgs-ova-photon-3-v1.23.8---vmware.3-tkg.1
  Storage Class:   wcpglobalstorageprofile
Status:
  Conditions:
    Last Transition Time:  2023-12-14T20:01:46Z
    Message:               failed to create or update VirtualMachine: admission webhook "default.validating.virtualmachine.v1alpha1.vmoperator.vmware.com" denied the request: spec.imageName: Invalid value: "tkgs-ova-photon-3-v1.23.8---vmware.3-tkg.1": field is immutable
    Reason:                VMCreationFailed
    Severity:              Warning
    Status:                False
    Type:                  Ready
    Last Transition Time:  2023-12-14T20:01:46Z
    Message:               failed to create or update VirtualMachine: admission webhook "default.validating.virtualmachine.v1alpha1.vmoperator.vmware.com" denied the request: spec.imageName: Invalid value: "tkgs-ova-photon-3-v1.23.8---vmware.3-tkg.1": field is immutable
    Reason:                VMCreationFailed
    Severity:              Warning
    Status:                False
    Type:                  VMProvisioned
  Vmstatus:                pending
```
This patch ensures the immutable fields on the VMop VM Spec do not get set if these  fields on the existing VM object are not empty.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2430

### CODE REVIEW

**Review by chrischdi — COMMENTED** [review]
First of all thanks for the PR!

This PR also contains some optimizations we might not want to cherry-pick.

Should we reduce this PR to the bare minimum required to solve the related issue and do the other changes on a separate PR?

That would come down to the changes in:
https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2566/files#diff-c2e693327f8831614b68401c025d9322280d92938f4a064bcd949656feb52770R179-R337

if I did not miss a point here.

File: pkg/services/vmoperator/vmoperator_suite_test.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package vmoperator
+package vmoperator_test
```
Comment (fabriziopandini) [inline pkg/services/vmoperator/vmoperator_suite_test.go:17]: q: why are moving this to a _test, AFAIK this is not necessary

File: pkg/services/vmoperator/vmopmachine_test.go [hunk 2]
```diff
@@ -373,7 +386,7 @@ var _ = Describe("VirtualMachine tests", func() {
 			requeue, err = vmService.ReconcileNormal(ctx, supervisorMachineContext)
 			verifyOutput(supervisorMachineContext)
 
-			Expect(vmopVM.Spec.ReadinessProbe.TCPSocket.Port.IntValue()).To(Equal(defaultAPIBindPort))
+			Expect(vmopVM.Spec.ReadinessProbe.TCPSocket.Port.IntValue()).To(Equal(6443))
```
Comment (fabriziopandini) [inline pkg/services/vmoperator/vmopmachine_test.go:389]: q why are we dropping the const?

File: pkg/services/vmoperator/vmopmachine_test.go [hunk 3]
```diff
@@ -513,7 +526,7 @@ var _ = Describe("VirtualMachine tests", func() {
 			Expect(vmopVM.Spec.Volumes).To(HaveLen(2))
 
 			for i, volume := range vsphereMachine.Spec.Volumes {
-				name := volumeName(vsphereMachine, volume)
+				name := fmt.Sprintf("%s-%s", vsphereMachine.Name, volume.Name)
```
Comment (fabriziopandini) [inline pkg/services/vmoperator/vmopmachine_test.go:529]: q why are we dropping the volumeName func?

**Review by fabriziopandini — COMMENTED** [review]
just a couple of nits from my side

File: pkg/services/vimmachine.go [hunk 4]
```diff
@@ -47,8 +47,31 @@ type VimMachineService struct {
 	Client client.Client
 }
 
+// GetMachinesInCluster returns a list of VSphereMachine objects belonging to the cluster.
+func (v *VimMachineService) GetMachinesInCluster(
```
[the remark this answers is not in the document]
Comment (srm09) [reply pkg/services/vimmachine.go:51]: Makes sense, since the client is a part of the `VimMachineService` and `VMopMachineService`, no use passing them in the function call. Missed that since most of it was copying over the function as is from the util package.

File: pkg/services/vimmachine.go [hunk 5]
```diff
@@ -26,7 +26,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
+	apitypes "k8s.io/apimachinery/pkg/types"
```
[the remark this answers is not in the document]
Comment (srm09) [reply pkg/services/vimmachine.go]: reverted back to the original state.

File: pkg/services/vmoperator/constants.go [hunk 6]
```diff
@@ -31,3 +31,30 @@ const (
 	// WorkerVMVMAntiAffinityTagValue is the value used for ProviderTagsAnnotationKey when the machine is a worker machine.
 	WorkerVMVMAntiAffinityTagValue = "WorkerVmVmAATag"
 )
+
+const (
+	// ClusterSelectorKey ...
```
[the remark this answers is not in the document]
Comment (srm09) [reply pkg/services/vmoperator/constants.go]: Reverted back to the original state of not being exported.

File: pkg/services/vmoperator/vmopmachine_test.go [hunk 7]
```diff
@@ -14,10 +14,11 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package vmoperator
+package vmoperator_test
```
[the remark this answers is not in the document]
Comment (srm09) [reply pkg/services/vmoperator/vmopmachine_test.go]: It was part of running tests in a separate package, but golang does not prescribe it as a convention. Removed this change as it was not relevant to the primary fix in the PR. We could do this in a follow up PR if need be.

File: pkg/services/vmoperator/vmoperator_suite_test.go (hunk 1 above)
Comment (srm09) [reply pkg/services/vmoperator/vmoperator_suite_test.go]: answered above, reverted.

File: pkg/services/vmoperator/vmopmachine_test.go (hunk 2 above)
Comment (srm09) [reply pkg/services/vmoperator/vmopmachine_test.go]: This was needed as part of moving the tests under a separate `_test` package. Reverted it since the change is not relevant to the PR.

File: pkg/services/vmoperator/vmopmachine_test.go (hunk 3 above)
Comment (srm09) [reply pkg/services/vmoperator/vmopmachine_test.go]: Same as above, reverted.

File: pkg/services/vmoperator/vmopmachine.go [hunk 8]
```diff
@@ -379,9 +415,9 @@ func (v *VmopMachineService) reconcileProviderID(ctx context.Context, supervisor
 	}
 }
 
-// getVirtualMachinesInCluster returns all VMOperator VirtualMachine objects in the current cluster.
-// First filter by clusterSelectorKey. If the result is empty, they fall back to legacyClusterSelectorKey.
-func (v *VmopMachineService) getVirtualMachinesInCluster(ctx context.Context, supervisorMachineCtx *vmware.SupervisorMachineContext) ([]*vmoprv1.VirtualMachine, error) {
+// GetVirtualMachinesInCluster returns all VMOperator VirtualMachine objects in the current cluster.
+// First filter by ClusterSelectorKey. If the result is empty, they fall back to LegacyClusterSelectorKey.
+func (v *VmopMachineService) GetVirtualMachinesInCluster(ctx context.Context, supervisorMachineCtx *vmware.SupervisorMachineContext) ([]*vmoprv1.VirtualMachine, error) {
```
Comment (chrischdi) [inline pkg/services/vmoperator/vmopmachine.go:420]: Let's revert this


```suggestion
// getVirtualMachinesInCluster returns all VMOperator VirtualMachine objects in the current cluster.
// First filter by clusterSelectorKey. If the result is empty, they fall back to legacyClusterSelectorKey.
func (v *VmopMachineService) getVirtualMachinesInCluster(ctx context.Context, supervisorMachineCtx *vmware.SupervisorMachineContext) ([]*vmoprv1.VirtualMachine, error) {
```

File: pkg/services/vmoperator/vmopmachine.go [hunk 9]
```diff
@@ -26,7 +26,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
+	apitypes "k8s.io/apimachinery/pkg/types"
```
Comment (chrischdi) [inline pkg/services/vmoperator/vmopmachine.go:29]:
```suggestion
	"k8s.io/apimachinery/pkg/types"
```

File: pkg/util/testutil.go [hunk 10]
```diff
@@ -37,7 +37,7 @@ import (
 
 const (
 	clusterKind          = "Cluster"
-	infraClusterKind     = "VSphereCluster"
+	InfraClusterKind     = "VSphereCluster"
```
Comment (chrischdi) [inline pkg/util/testutil.go:40]: Let's revert this
```suggestion
	infraClusterKind     = "VSphereCluster"
```

File: controllers/vspheremachine_controller.go [hunk 11]
```diff
@@ -108,7 +108,7 @@ func AddMachineControllerToManager(ctx context.Context, controllerManagerContext
 			).
 			Watches(
 				&clusterv1.Cluster{},
-				handler.EnqueueRequestsFromMapFunc(r.clusterToVMwareMachines),
+				handler.EnqueueRequestsFromMapFunc(r.enqueueClusterToMachineRequests),
```
Comment (chrischdi) [inline controllers/vspheremachine_controller.go:111]: Sorry for stressing this and causing some more work, let's revert the changes around this for this PR and move them to a separate PR. (Includes reverting the additional new / moved functions)

I'm happy to help out if you don't find the time for it.

File: controllers/vspheremachine_controller.go [hunk 12]
```diff
@@ -161,7 +161,7 @@ func AddMachineControllerToManager(ctx context.Context, controllerManagerContext
 		).
 		Watches(
 			&clusterv1.Cluster{},
-			handler.EnqueueRequestsFromMapFunc(r.clusterToVSphereMachines),
+			handler.EnqueueRequestsFromMapFunc(r.enqueueClusterToMachineRequests),
```
Comment (chrischdi) [inline controllers/vspheremachine_controller.go:164]: Sorry for stressing this and causing some more work, let's revert the changes around this for this PR and move them to a separate PR.

**Review by chrischdi — COMMENTED** [review]
Some more nits + please move the predicate relevant changes to a separate PR :-)

**Comment by srm09** [comment]
Reverted all the irrelavant changes so that this one just handles skipping immutable fields

**Review by chrischdi — COMMENTED** [review]
/lgtm
/approve

File: pkg/services/vmoperator/vmopmachine.go [hunk 13]
```diff
@@ -287,17 +293,25 @@ func (v *VmopMachineService) reconcileVMOperatorVM(ctx context.Context, supervis
 		// Define a new VM Operator virtual machine.
 		// NOTE: Set field-by-field in order to preserve changes made directly
 		//  to the VirtualMachine spec by other sources (e.g. the cloud provider)
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply pkg/services/vmoperator/vmopmachine.go:295]: Accordung vm-operator code it is als MinHardwareVersion:
https://github.com/vmware-tanzu/vm-operator/blob/main/webhooks/virtualmachine/v1alpha1/validation/virtualmachine_validator.go#L143-L149

but great catch, we should also have the same behaviour for `ResourcePolicyName`!

Also +1 for making these immutable on CAPV side. We will have to introduce webhooks for vm-operator mode capv first!

[the remark this answers is not in the document]
Comment (chrischdi) [reply pkg/services/vmoperator/vmopmachine.go:295]: created #2595 to track this validating webhook part.

[the remark this answers is not in the document]
Comment (sbueringer) [reply pkg/services/vmoperator/vmopmachine.go:295]: Thank you!
